src/datadog/config_manager.cpp

```diff
@@ -1,80 +1,164 @@
+#include "parse_util.h"
 #include "trace_sampler.h"
 
 namespace datadog {
 namespace tracing {
+namespace {
+
+std::string join(const std::vector<StringView>& values,
+                 const char* const separator) {
+  if (values.empty()) return {};
+  auto it = values.cbegin();
+
+  std::string res{*it};
+  for (++it; it != values.cend(); ++it) {
+    res += separator;
+    append(res, *it);
+  }
+
+  return res;
+}
+
+// TODO: use `to_chars`
+std::string to_string(double d) {
+  std::stringstream stream;
+  stream << std::fixed << std::setprecision(1) << d;
+  return stream.str();
+}
+
+}  // namespace
 
 ConfigManager::ConfigManager(const FinalizedTracerConfig& config)
     : clock_(config.clock),
-      default_trace_sampler_(
+      default_config_metadata_(config.metadata),
+      trace_sampler_(
           std::make_shared<TraceSampler>(config.trace_sampler, clock_)),
-      current_trace_sampler_(default_trace_sampler_),
-      default_span_defaults_(std::make_shared<SpanDefaults>(config.defaults)),
-      current_span_defaults_(default_span_defaults_),
-      default_report_traces_(config.report_traces),
-      current_report_traces_(default_report_traces_) {}
+      span_defaults_(std::make_shared<SpanDefaults>(config.defaults)),
+      report_traces_(config.report_traces) {}
 
 std::shared_ptr<TraceSampler> ConfigManager::trace_sampler() {
   std::lock_guard<std::mutex> lock(mutex_);
-  return current_trace_sampler_;
+  return trace_sampler_;
 }
 
 std::shared_ptr<const SpanDefaults> ConfigManager::span_defaults() {
   std::lock_guard<std::mutex> lock(mutex_);
-  return current_span_defaults_;
+  return span_defaults_;
 }
 
 bool ConfigManager::report_traces() {
   std::lock_guard<std::mutex> lock(mutex_);
-  return current_report_traces_;
+  return report_traces_;
 }
 
-void ConfigManager::update(const ConfigUpdate& conf) {
+std::vector<ConfigMetadata> ConfigManager::update(const ConfigUpdate& conf) {
+  std::vector<ConfigMetadata> telemetry_conf;
+
   std::lock_guard<std::mutex> lock(mutex_);
 
-  if (conf.trace_sampler) {
-    if (auto finalized_trace_sampler_cfg =
-            finalize_config(*conf.trace_sampler)) {
-      current_trace_sampler_ =
-          std::make_shared<TraceSampler>(*finalized_trace_sampler_cfg, clock_);
-    } else {
-      // TODO: report error
+  if (conf.trace_sampling_rate) {
+    ConfigMetadata trace_sampling_metadata(
+        ConfigName::TRACE_SAMPLING_RATE, to_string(*conf.trace_sampling_rate),
+        ConfigMetadata::Origin::REMOTE_CONFIG);
+
+    TraceSamplerConfig trace_sampler_cfg;
+    trace_sampler_cfg.sample_rate = *conf.trace_sampling_rate;
+
+    auto finalized_trace_sampler_cfg = finalize_config(trace_sampler_cfg);
+    if (auto error = finalized_trace_sampler_cfg.if_error()) {
+      trace_sampling_metadata.error = *error;
+    }
+
+    auto trace_sampler =
+        std::make_shared<TraceSampler>(*finalized_trace_sampler_cfg, clock_);
+
+    if (trace_sampler != trace_sampler_.get()) {
```

Comment:
This is comparing `shared_ptr`s.

Reply:
addressed in e2bb8f465b2245055cb150d6d5bc8bcf69211c65